linkerd/app/outbound/src/metrics/transport.rs

```diff
@@ -0,0 +1,271 @@
+            Self::InvalidPolicy => write!(f, "invalid_policy"),
+            Self::Unexpected => write!(f, "unexpected"),
+        }
+    }
+}
+
+// === impl ConnectionsClosedLabels ===
+
+impl<L> EncodeLabelSetMut for ConnectionsClosedLabels<L>
+where
+    L: Clone + Hash + Eq + EncodeLabelSetMut + Debug + Send + Sync + 'static,
+{
+    fn encode_label_set(&self, enc: &mut LabelSetEncoder<'_>) -> std::fmt::Result {
+        self.labels.encode_label_set(enc)?;
+        match self.error {
+            Some(error) => ("error", error.to_string()).encode(enc.encode_label())?,
+            None => ("error", "").encode(enc.encode_label())?,
+        }
+
+        Ok(())
+    }
+}
+
+impl<L> EncodeLabelSet for ConnectionsClosedLabels<L>
+where
+    L: Clone + Hash + Eq + EncodeLabelSetMut + Debug + Send + Sync + 'static,
+{
+    fn encode(&self, mut enc: LabelSetEncoder<'_>) -> std::fmt::Result {
+        self.encode_label_set(&mut enc)
+    }
+}
+
+// === impl NewTransportRouteMetrics ===
+
+impl<N, L: Clone> NewTransportRouteMetrics<N, L> {
+    pub fn layer(
+        family: TransportRouteMetricsFamily<L>,
+    ) -> impl layer::Layer<N, Service = Self> + Clone {
+        layer::mk(move |inner| Self {
+            inner,
+            family: family.clone(),
+        })
+    }
+}
+
+impl<T, N, L> NewService<T> for NewTransportRouteMetrics<N, L>
+where
+    N: Clone,
+    L: Clone + Hash + Eq + EncodeLabelSetMut + Debug + Send + Sync + 'static,
+    T: Param<L>,
+{
+    type Service = TransportRouteMetricsService<T, N>;
+
+    fn new_service(&self, target: T) -> Self::Service {
+        let labels: L = target.param();
+        let metrics = self.family.metrics(labels);
+        TransportRouteMetricsService::new(target, self.inner.clone(), metrics)
+    }
+}
+
+// === impl TransportRouteMetricsService ===
+
+impl<T, N> TransportRouteMetricsService<T, N> {
+    fn new(target: T, inner: N, metrics: TransportRouteMetrics) -> Self {
+        Self {
+            target,
+            inner,
+            metrics,
+        }
+    }
+}
+
+impl<T, I, N, S> Service<I> for TransportRouteMetricsService<T, N>
+where
+    T: Clone + Send + Sync + 'static,
+    I: io::AsyncRead + io::AsyncWrite + Debug + Send + Unpin + 'static,
+    N: NewService<T, Service = S> + Clone + Send + 'static,
+    S: Service<I> + Send,
+    S::Error: Into<Error>,
+    S::Future: Send,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = Pin<Box<dyn Future<Output = Result<S::Response, Error>> + Send + 'static>>;
+
+    #[inline]
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, io: I) -> Self::Future {
+        let target = self.target.clone();
+        let new_accept = self.inner.clone();
+        let metrics = self.metrics.clone();
+
+        Box::pin(async move {
+            let svc = new_accept.new_service(target);
```

Comment:
Can't this happen in the NewService impl so that this middleware can properly proxy poll_ready()?

Reply:
Will give it a try

Reply:
Addressed in https://github.com/linkerd/linkerd2-proxy/pull/3355/commits/2549251dd15c71a8e64cb0bf9673e2515669ea16